src/coreclr/gc/gc.cpp

```diff
@@ -11541,6 +11532,477 @@ void gc_heap::rearrange_heap_segments(BOOL compacting)
+            if (region_size == LARGE_REGION_SIZE * 2)
+            {
+                // we won't find a smaller one on this list
+                assert (region == smallest_region);
+                break;
+            }
+        }
+    }
+
+    if (smallest_region != nullptr)
+    {
+        unlink_region (smallest_region);
+        dprintf(REGIONS_LOG, ("get %Ix-%Ix-%Ix",
+            heap_segment_mem(smallest_region), heap_segment_committed(smallest_region), heap_segment_used(smallest_region)));
+    }
+
+    return smallest_region;
+}
+
+void region_free_list::transfer_regions (region_free_list* from)
+{
+    if (from->num_free_regions == 0)
+    {
+        // the from list is empty
+        assert (from->size_free_regions == 0);
+        assert (from->size_committed_in_free_regions == 0);
+        assert (from->head_free_region == nullptr);
+        assert (from->tail_free_region == nullptr);
+        return;
+    }
+    if (num_free_regions == 0)
+    {
+        // this list is empty
+        assert (size_free_regions == 0);
+        assert (size_committed_in_free_regions == 0);
+        assert (head_free_region == nullptr);
+        assert (tail_free_region == nullptr);
+
+        head_free_region = from->head_free_region;
+        tail_free_region = from->tail_free_region;
+    }
+    else
+    {
+        // both free lists are non-empty
+        assert (from->size_free_regions != 0);
+        assert (from->size_committed_in_free_regions != 0);
+        assert (from->head_free_region != nullptr);
+        assert (from->tail_free_region != nullptr);
+
+        assert (size_free_regions != 0);
+        assert (size_committed_in_free_regions != 0);
+        assert (head_free_region != nullptr);
+        assert (tail_free_region != nullptr);
+
+        // attach the from list at the tail
+        heap_segment* this_tail = tail_free_region;
+        heap_segment* from_head = from->head_free_region;
+
+        assert (heap_segment_next (this_tail) == nullptr);
+        assert (heap_segment_prev_free_region (from_head) == nullptr);
+
+        heap_segment_next (this_tail) = from_head;
+        heap_segment_prev_free_region (from_head) = this_tail;
+
+        tail_free_region = from->tail_free_region;
+
+    }
+
+    for (heap_segment* region = from->head_free_region; region != nullptr; region = heap_segment_next (region))
+    {
+        heap_segment_containing_free_list (region) = this;
+    }
+
+    num_free_regions += from->num_free_regions;
+    num_free_regions_added += from->num_free_regions;
+    size_free_regions += from->size_free_regions;
+    size_committed_in_free_regions += from->size_committed_in_free_regions;
+
+    from->num_free_regions_removed += from->num_free_regions;
+    from->num_free_regions = 0;
+    from->size_free_regions = 0;
+    from->size_committed_in_free_regions = 0;
+
+    from->head_free_region = nullptr;
+    from->tail_free_region = nullptr;
+
```

Comment:
nit - could we have an init method that this can call instead of having to initialize each field individually? you could do the assert on the from->num_free_regions first, then init the whole thing to 0

Reply:
I think we wouldn't want to clear the num_free_regions_added and num_free_regions_removed fields, so I introduced a reset method that is called here.